### PR TITLE
OrleansSQLUtils port to .Net Standard

### PR DESCRIPTION
**Pull request by galvesribeiro**
As part of .Net Core port (#2143) at the initial CRAWL phase (#2145) we need to port all the providers officially supported by Orleans and that includes OrleansSQLUtils.

In this PR, the providers for relational databases (SQL Server, PostgreSQL, MySQL, SQLite, Oracle), are ported AS IS in order to facilitate the merge. The only major change was in `DbFactoryProvider` class because DbFactories in ADO.Net for .Net Core doesn't exist. Because of that, I had to change its internal implementation. Also, as part of the test cleanup effort on vNext solution, all relational tests were moved to is own project. 

This PR try to follow the work of @xiazen on OrleansAzureUtils port to minimize reviews efforts.

### CODE REVIEW

File: src/Orleans.sln [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 14
-VisualStudioVersion = 14.0.25420.1
```
Comment (galvesribeiro) [inline src/Orleans.sln]: I have only VS2017 now. Also that has no effect on the generated files, it is just the VS Version. Honestly, it shouldn't be there... Everyone that has an update of VS and open the solution will have this value changed. Hard to keep older ones...

File: src/OrleansSQLUtils/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -21,4 +22,7 @@
 [assembly: InternalsVisibleTo("UnitTests")]
 [assembly: InternalsVisibleTo("TesterInternal")]
 [assembly: InternalsVisibleTo("UnitTestGrains")]
+[assembly: InternalsVisibleTo("Orleans.NonSiloTests")]
```
Comment (xiazen) [inline src/OrleansSQLUtils/Properties/AssemblyInfo.cs]: Yeah I need to clean up NonSiloTests... did couple things wrong there. So questions to other 
[assembly: InternalsVisibleTo("Tester.SQLUtils")] , you added those because it is needed in order to compile or because AzureUtils PR did it? 
Because those two are not completely equal.

Comment (xiazen) [inline src/OrleansSQLUtils/Properties/AssemblyInfo.cs]: cannot check it for you since main solution doesn't compile, for the reason in comments below

File: src/OrleansSQLUtils/Storage/DbConnectionFactory.cs [hunk 3]
```diff
@@ -1,39 +1,62 @@
...
         {
             if (String.IsNullOrWhiteSpace(invariantName))
             {
                 throw new ArgumentNullException(nameof(invariantName));
             }
 
-            // Seeks for database provider factory classes from GAC or as indicated by
-            // the configuration file, see at <see href="https://msdn.microsoft.com/en-us/library/dd0w4a2z%28v=vs.110%29.aspx">Obtaining a DbProviderFactory</see>.       
+            if (!providerFactoryTypeMap.TryGetValue(invariantName, out Tuple<string, string> providerFactoryDefinition))
+                throw new InvalidOperationException($"Database provider factory with '{invariantName}' invariant name not supported.");
```
Comment (xiazen) [inline src/OrleansSQLUtils/Storage/DbConnectionFactory.cs]: should be something like below
```
            Tuple<string, string> providerFactoryDefinition;
            if (!providerFactoryTypeMap.TryGetValue(invariantName, out providerFactoryDefinition))
```
to make it compile

Comment (xiazen) [inline src/OrleansSQLUtils/Storage/DbConnectionFactory.cs]: You need to declare providerFactoryDefinition first

File: src/OrleansSQLUtils/Properties/AssemblyInfo.cs (hunk 2 above)
Comment (galvesribeiro) [inline src/OrleansSQLUtils/Properties/AssemblyInfo.cs]: I added to compile. Provider's test still require internals at some point. We discussed that we need to create a "Provider Test Kit" package so we can have it all removed. For now we need keep it.

File: src/OrleansSQLUtils/Storage/DbConnectionFactory.cs (hunk 3 above)
Comment (xiazen) [inline src/OrleansSQLUtils/Storage/DbConnectionFactory.cs]: not compiling on my machine. maybe I'm crazy, but I did build netstandard after git clean -xdf

Comment (xiazen) [inline src/OrleansSQLUtils/Storage/DbConnectionFactory.cs]: I believe Jenkins cannot compile it too. 

> "D:\j\workspace\netfx_prtestb457c4cd\src\OrleansSQLUtils\OrleansSQLUtils.csproj" (default target) (9) ->
       (CoreCompile target) -> 
         Storage\DbConnectionFactory.cs(33,78): error CS1525: Invalid expression term 'string' [D:\j\workspace\netfx_prtestb457c4cd\src\OrleansSQLUtils\OrleansSQLUtils.csproj]
         Storage\DbConnectionFactory.cs(33,86): error CS1525: Invalid expression term 'string' [D:\j\workspace\netfx_prtestb457c4cd\src\OrleansSQLUtils\OrleansSQLUtils.csproj]

I got the same error when compile locally

**Comment by xiazen** [comment]
Your netstandard build failed on a lot of binding redirects issues.

**Comment by xiazen** [comment]
if you are working on change our test script to use dotnet test, we can merge that PR first and then this one. No rush ;)

**Comment by galvesribeiro** [comment]
I'm not changing the test script neither the project to `dotnet test`.

File: test/TesterSQLUtils/TesterSQLUtils.csproj [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    <ProjectReference Include="..\TestInternalGrains\TestInternalGrains.csproj">
+      <Project>{4682abc1-1a51-4794-8012-da8e59ebc72e}</Project>
+      <Name>TestInternalGrains</Name>
+    </ProjectReference>
+  </ItemGroup>
+  <ItemGroup>
+    <Content Include="..\..\src\OrleansSQLUtils\CreateOrleansTables_MySql.sql">
+      <Link>CreateOrleansTables_MySql.sql</Link>
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
```
Comment (xiazen) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: PreserveNewest?

File: test/TesterSQLUtils/TesterSQLUtils.csproj [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+      <Link>CreateOrleansTables_PostgreSql.sql</Link>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </Content>
+    <Content Include="..\..\src\OrleansSQLUtils\CreateOrleansTables_SqlServer.sql">
+      <Link>CreateOrleansTables_SqlServer.sql</Link>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </Content>
+    <Content Include="Data\TestDb.mdf">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
```
Comment (xiazen) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: same here

File: test/TesterSQLUtils/TesterSQLUtils.csproj [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    <Content Include="Data\TestDb_log.ldf">
+      <DependentUpon>TestDb.mdf</DependentUpon>
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
```
Comment (xiazen) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: here

File: test/TesterSQLUtils/TesterSQLUtils.csproj [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <BootstrapperPackage Include="Microsoft.Net.Framework.3.5.SP1">
+      <Visible>False</Visible>
+      <ProductName>.NET Framework 3.5 SP1</ProductName>
+      <Install>false</Install>
+    </BootstrapperPackage>
+  </ItemGroup>
```
Comment (xiazen) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: what's this part for? other test projects doesn't have it

File: vNext/src/NuGet/Microsoft.Orleans.OrleansSQLUtils.nuspec [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    <tags>Orleans Cloud-Computing Actor-Model Actors Distributed-Systems C# .NET</tags>
+    <dependencies>
+      <group targetFramework=".NETStandard1.5">
+        <dependency id="Microsoft.Orleans.Core" version="$version$" />
+        <dependency id="Microsoft.Orleans.OrleansRuntime" version="$version$" />
+        <dependency id="Microsoft.Orleans.OrleansProviders" version="$version$" />
+        <dependency id="System.Data.Common" version="4.3.0" />
+        <dependency id="System.Reflection" version="4.3.0" />
+        <dependency id="System.Xml.XmlSerializer" version="4.3.0" />
+      </group>
```
Comment (xiazen) [inline vNext/src/NuGet/Microsoft.Orleans.OrleansSQLUtils.nuspec]: I noticed that in src/Nuget/Microsoft.Orleans.OrleansSQLUtils.nuspec  it only have dependency on Orleans.Core and so does OrleansSQLUtils csproj in main solution. 

I'm skeptical that when it is moved to vNext, it needs OrleansRuntime and OrleansProviders as dependency

File: vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj [hunk 9]
```diff
@@ -0,0 +1,56 @@
+﻿<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" />
+  <PropertyGroup Label="Configuration">
+    <DefineConstants>NETSTANDARD;NETSTANDARD_TODO</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup>
+    <TargetFramework>netstandard1.5</TargetFramework>
+    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
+    <PackageTargetFallback>$(PackageTargetFallback);portable-net45+win8+wpa81+wp8</PackageTargetFallback>
```
Comment (xiazen) [inline vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj]: you don't necessary need this. AzureUtils need this is because Windows.AzureStorage package need this. Couple of AzureStorage package's dependency lib doesn't support netstandard yet. So need this as a fallback to load those libs. Took me a long time nicely fix it. 

Not sure SQLUtils need this. try remove it and to see if it builds

File: test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 4 above)
Comment (galvesribeiro) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: Done

File: test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 5 above)
Comment (galvesribeiro) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: Done

File: test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 6 above)
Comment (galvesribeiro) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: Done

File: test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 7 above)
Comment (galvesribeiro) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: I don't know how it ended up there... But in the current file I have here, it isn't here...

File: vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj (hunk 9 above)
Comment (galvesribeiro) [inline vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj]:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB4057	The target "ResolvePackageDependenciesForBuild" does not exist in the project.	ClientGenerator	C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets	1871	
```
Right after remove it

File: vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <None Update="..\..\..\src\OrleansSQLUtils\CreateOrleansTables_MySql.sql">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="..\..\..\src\OrleansSQLUtils\CreateOrleansTables_PostgreSql.sql">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="..\..\..\src\OrleansSQLUtils\CreateOrleansTables_SqlServer.sql">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
```
Comment (xiazen) [inline vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj]: what is the `Update ` keyword? Isn't it should be `Include` ? 
this is causing this project not be able to load in my vs17.

Comment (galvesribeiro) [inline vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj]: In line 17 the .sql files are globbed and the Update keywork is to get some previously defined item and update one of its properties. VS Generate it is you right click -> properties on a globbed file and you set for example Copyt to output in it, which is precisely the case.

Idk why your VS isn't working. I'm using the public RC, you are probably with an internal build that has it different. 

In all cases, I'll remove the globb on line 17 and change this to include, will do the same effect and unblock you...

File: vNext/test/TesterSQLUtils/TesterSQLUtils.csproj [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    <PackageReference Include="MySql.Data">
+      <Version>7.0.6-IR31</Version>
+    </PackageReference>
+    <PackageReference Include="MySql.Data">
+      <Version>4.3.0</Version>
+    </PackageReference>
+    <PackageReference Include="System.Runtime">
+      <Version>4.3.0</Version>
+    </PackageReference>
+  </ItemGroup>
```
Comment (xiazen) [inline vNext/test/TesterSQLUtils/TesterSQLUtils.csproj]: I believe you just need those in app.config to be able to do binary redirect. not necessary to include all of them as a PackageReference. Many of them are not direct dependency package right?  like `System.Security.Cryptography.Primitives` etc

File: vNext/test/TesterSQLUtils/TesterSQLUtils.csproj [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+    <ProjectReference Include="..\Tester\Tester.csproj" />
+    <ProjectReference Include="..\TestExtensions\TestExtensions.csproj" />
+    <ProjectReference Include="..\TestGrainInterfaces\TestGrainInterfaces.csproj" />
+    <ProjectReference Include="..\TestGrains\TestGrains.csproj" />
+    <ProjectReference Include="..\TestInterfaces\TestInterfaces.csproj" />
+    <ProjectReference Include="..\TestInternalGrainInterface\TestInternalGrainInterfaces.csproj" />
+    <ProjectReference Include="..\TestInternalGrains\TestInternalGrains.csproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="..\..\..\test\TesterSQLUtils\RelationalUtilities\MySqlStorageForTesting.cs">
```
Comment (xiazen) [inline vNext/test/TesterSQLUtils/TesterSQLUtils.csproj]: you don't need to include those cs one by one. You can do a wild card   <Compile Include="**\*.cs" /> and exclude those you don't want to . if the included file in TesterSQLUtils in VNext the same as in Main? if not, which are not included ?

File: vNext/test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 11 above)
Comment (galvesribeiro) [inline vNext/test/TesterSQLUtils/TesterSQLUtils.csproj]: They are pulled automatically when you are in a .Net Core app, since it is a class library targeting netstandard and we are adding project reference and not the OrleansSQLUtils nuget itself, it needs to be manually referenced by the test project. In other words, it is an indirect dependency that must be present at runtime.

File: vNext/test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 12 above)
Comment (galvesribeiro) [inline vNext/test/TesterSQLUtils/TesterSQLUtils.csproj]: I followed the pattern of Azure Util tests and, I'm not sure that globbing is available for net452 projects... Need to confirm that...

File: vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj (hunk 10 above)
Comment (xiazen) [inline vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj]: What's the `Update ` keyword, isn't it should be `Include `?  
what does it do? it is a new feature in vs17 I think

Comment (galvesribeiro) [inline vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj]: See -> https://github.com/dotnet/orleans/pull/2470#discussion_r90583128

Comment (xiazen) [inline vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj]: I used vs 15 to load it, that was my fault. So I deleted the comment.

File: test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 7 above)
Comment (galvesribeiro) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: Probably VS2017 magic for something... Now it isn't here on my current project...

File: vNext/test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 12 above)
Comment (xiazen) [inline vNext/test/TesterSQLUtils/TesterSQLUtils.csproj]: ah yeah that might not work. forgot it is in 462. but good to confirm

File: test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 7 above)
Comment (xiazen) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: remove it to see? maybe it is not needed?

File: vNext/src/NuGet/Microsoft.Orleans.OrleansSQLUtils.nuspec (hunk 8 above)
Comment (xiazen) [inline vNext/src/NuGet/Microsoft.Orleans.OrleansSQLUtils.nuspec]: Yeah AzureUtils has dependency on those two . you can check it in Main solution

File: test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 7 above)
Comment (galvesribeiro) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: Already did that... Dunno why it still here in the review...

File: vNext/src/NuGet/Microsoft.Orleans.OrleansSQLUtils.nuspec (hunk 8 above)
Comment (galvesribeiro) [inline vNext/src/NuGet/Microsoft.Orleans.OrleansSQLUtils.nuspec]: Done

File: src/OrleansSQLUtils/Storage/DbConnectionFactory.cs (hunk 3 above)
Comment (galvesribeiro) [inline src/OrleansSQLUtils/Storage/DbConnectionFactory.cs]: Looks like it has an old compiler... Reverting to "traditional" way...

File: vNext/test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 11 above)
Comment (jdom) [inline vNext/test/TesterSQLUtils/TesterSQLUtils.csproj]: Hmmm, on a quick glance, I think Xiao might be right... at least the cryptography one looks like a transitive dependency. This code in particular is not using, just one of the packages that you do use is.

**Comment by galvesribeiro** [comment]
@xiazen rebased and removed the dependencies as requested

File: src/Orleans.sln (hunk 1 above)
Comment (xiazen) [inline src/Orleans.sln]: revert this. it is just unnecessary diff, may cause unnecessary merge conflict to other member

Comment (galvesribeiro) [inline src/Orleans.sln]: ???

Comment (xiazen) [inline src/Orleans.sln]: what I mean is this diff is not needed for your change, normally we don't include diff which is not necessary,  for reasons like avoiding merge conflicts

File: test/TesterSQLUtils/TesterSQLUtils.csproj (hunk 7 above)
Comment (xiazen) [inline test/TesterSQLUtils/TesterSQLUtils.csproj]: are you sure? I still see it in the diff.

File: src/Orleans.sln (hunk 1 above)
Comment (galvesribeiro) [inline src/Orleans.sln]: @xiazen... Ok, will not arg, again... I wonder if you will hold this change to everyone every time they open Visual Studio with newer version, including core team... Anyway, consider done for the sake to finish this once and for all... We are looking for a problem that doesn't exist...

Comment (xiazen) [inline src/Orleans.sln]: just for the record. I personally do it. And I asked another person from the core team too and he does it. I didn't do it to you personally. So don't take it personal. 

avoid unnecessary change is the first thing I learnt in my job. If I did this, any team I worked before for would be against. 

And it is not hard to revert this at all.

Comment (xiazen) [inline src/Orleans.sln]: this maybe inconvenient to you. I wonder if there's a way to turn it off. 

But avoiding unnecessary change is first thing I learnt as a SDE. And this is a common standard based on my pass experience with couple teams in Amazon. it decrease work for your reviewer and people in the future who will be looking at this commit. 

Don't take it personal. I'm just apply the same standard to you as to other people and as other people to me. And just for the record, I asked a core team member, he does it too.

Comment (xiazen) [inline src/Orleans.sln]: And this is not the main thing which holds this PR off at all. I think we talked about running the sqlutils tests before merging this.

Comment (xiazen) [inline src/Orleans.sln]: You can assign the people who runs the sqlutils test regularly to this PR. They are the people who care about the quality of sqlutils most. And they will be able to help you.

File: vNext/src/NuGet/Microsoft.Orleans.OrleansSQLUtils.nuspec (hunk 8 above)
Comment (xiazen) [inline vNext/src/NuGet/Microsoft.Orleans.OrleansSQLUtils.nuspec]: dependency of OrleansRuntime and OrleansProviders is still here. make sure you commit the change you did.

File: vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    </PackageReference>
+    <PackageReference Include="System.Reflection">
+      <Version>4.3.0</Version>
+    </PackageReference>
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\Orleans\Orleans.csproj" />
+    <ProjectReference Include="..\OrleansRuntime\OrleansRuntime.csproj" />
+    <ProjectReference Include="..\OrleansProviders\OrleansProviders.csproj" />
+  </ItemGroup>
```
Comment (xiazen) [inline vNext/src/OrleansSQLUtils/OrleansSQLUtils.csproj]: I believe you don't need OrleansRuntime and OrleansProviders as dependency?

File: src/Orleans.sln (hunk 1 above)
Comment (galvesribeiro) [inline src/Orleans.sln]: Again, I'll not arg with you anymore since you have the "merge button power" and other people is waiting to have it merged and they are the ones I'm caring about with this work... 

I'll check every single merge from now on and see whether everyone will always keep it stuck on this version... Let's hope you are right... It just doesn't make sense...

Regarding the tests, you as an SDE should know that makes no sense to run or change a test that is not going to bring any benefits since there is no **fork** paths on this code for diff databases and the .sql scripts (which is the only thing that contain database-specific code) doesn't change, **AT ALL**. **All** the guys that created those providers already clearly stated that on Gitter. Anyone who read the code and properly review the changes will clearly note that only run the SQL Server tests will be enough to review/cover this code.

As I said on another comment, I'll run the tests on docker containers for MySQL and Postgre **just to make you happy** and will post here the results, because noone agree with your complains and arguments in that. 

There is nothing personal here. It is just very upset to have no sense complains about something that doesn't change and is being used forever by multiple people. If you want to really properly test every single case, change your CI environment on Jenkins and make all those tests BVT so it can run every single time a PR is created otherwise, we are just wasting precious time that should be focused on other areas that needs attention.

**Comment by galvesribeiro** [comment]
Ok... As requested, the tests on the other db engines were executed using containers:

image

image

As expected (and I repeated why over and over again), all tests pass. The remaining feedback was also addressed. 

I hope the reviewers are happy now. Please let me know if you need anything else.

Thank you.

**Comment by galvesribeiro** [comment]
Last minute change... After speak with @shayhatsor, the `GatewaySelect` test is unecessary and old since it is already implemented in this test https://github.com/dotnet/orleans/blob/master/test/TesterInternal/MembershipTests/MembershipTableTestsBase.cs#L87

**Comment by shayhatsor** [comment]
@galvesribeiro, Great Job !

**Comment by xiazen** [comment]
Thanks for testing it !

**Comment by sergeybykov** [comment]
@galvesribeiro, thank you so much!
